### PR TITLE
Render+geo

### PR DESCRIPTION
**Pull request by houguohua**

### CODE REVIEW

File: cocos/3d/C3DAABB.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	~C3DAABB();
+
+	/**
+	* Gets the center point of the bounding box.
+	*/
+	Vector3 getCenter();
+
+	Vector3 getSize();
+
+	float getLength();
```
Comment (minggo) [inline cocos/3d/C3DAABB.h:48]: Add `const` for `getXXX` functions?

File: cocos/3d/C3DNode.cpp [hunk 2]
```diff
@@ -0,0 +1,815 @@
...
+
+#include "C3DAABB.h"
+#include "C3DOBB.h"
+
+
+namespace cocos2d
+{
+
+C3DNode::C3DNode()   
+      : _scene(NULL),_parent(NULL),_active(true),
```
Comment (minggo) [inline cocos/3d/C3DNode.cpp:17]: NULL -> nullptr

File: cocos/3d/C3DNode.cpp [hunk 3]
```diff
@@ -0,0 +1,815 @@
...
+    }
+    
+    SAFE_DELETE(_bb);
+    SAFE_DELETE(_bbOrigin);
+    
+}
+
+C3DNode* C3DNode::create(const char* id)
+{
+    C3DNode* pRet = new C3DNode(id);
```
Comment (minggo) [inline cocos/3d/C3DNode.cpp:78]: Hungarian style.

File: cocos/3d/C3DNode.h [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+#include "C3DAABB.h"
+
+namespace cocos2d
+{
+
+    // C3DNode dirty flags
+#define NODE_DIRTY_WORLD 1
+#define NODE_DIRTY_BOUNDS_AABB 2
+#define NODE_DIRTY_BOUNDS_OBB 4
+#define NODE_DIRTY_ALL (NODE_DIRTY_WORLD | NODE_DIRTY_BOUNDS_AABB | NODE_DIRTY_BOUNDS_OBB)
```
Comment (minggo) [inline cocos/3d/C3DNode.h:19]: How about use const static member variable instead?

File: cocos/3d/C3DNode.h [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+class C3DCamera;
+class C3DViewport;
+class C3DMeshSkin;
+class C3DAABB;
+class C3DOBB;
+
+/**
+ * Defines a basic hierachial structure of transformation spaces.
+ */
+class C3DNode : public Transform, public virtual cocos2d::CCObject
```
Comment (minggo) [inline cocos/3d/C3DNode.h:34]: It is better not use multiple inheritance, use composition instead.

File: cocos/3d/C3DNode.h [hunk 6]
```diff
@@ -0,0 +1,483 @@
...
+         * @param transform The Transform object that was changed.
+         * @param cookie Cookie value that was specified when the listener was registered.
+         */
+        virtual void transformChanged(Transform* transform) = 0;
+    };
+
+    /**
+     * Defines the types of nodes.
+     */
+    enum Type
```
Comment (minggo) [inline cocos/3d/C3DNode.h:71]: Use enum class instead?

File: cocos/3d/C3DNode.h [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+
+    /**
+     *  Set active
+     */
+    virtual void active(bool active);
+
+    /**
+     *  Checks active
+     */
+    virtual bool active();
```
Comment (minggo) [inline cocos/3d/C3DNode.h:374]: active -> isActive?

File: cocos/3d/C3DNode.h [hunk 8]
```diff
@@ -0,0 +1,483 @@
...
+     * Returns the view-space translation vector of the currently active camera for this node's scene.
+     *
+     * @return The translation vector of the scene's active camera, in view-space.
+     */
+    const     Vector3& getActiveCameraTranslationView();    
+
+    /**
+     * Called when this Node's transform changes.
+     */
+    void transformChanged();
```
Comment (minggo) [inline cocos/3d/C3DNode.h:340]: transformChanged -> onTransformChanged?

File: cocos/3d/C3DNode.h [hunk 9]
```diff
@@ -0,0 +1,483 @@
...
+    
+protected:
+
+    C3DScene* _scene;
+    
+    std::string _id;
+
+    C3DNode* _parent;
+
+    std::vector< C3DNode * >  _children;  // Child nodes
```
Comment (minggo) [inline cocos/3d/C3DNode.h:445]: Now we can use `cocos2d::Vector<>` instead. It will handle reference correctly.

File: cocos/3d/C3DNode.h [hunk 10]
```diff
@@ -0,0 +1,483 @@
...
+    std::string _id;
+
+    C3DNode* _parent;
+
+    std::vector< C3DNode * >  _children;  // Child nodes    
+
+    /** 
+     * List of TransformListener's on the Transform.
+    */
+    std::list<Listener*>* _listeners;
```
Comment (minggo) [inline cocos/3d/C3DNode.h:450]: Can reuse `EventDispatcher`?

File: cocos/3d/C3DViewport.cpp [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+C3DViewport& C3DViewport::operator==(const C3DViewport& viewport)
+{
+	this->x = viewport.x;
+	this->y = viewport.y;
+	this->width = viewport.width;
+	this->height = viewport.height;
+
+	return *this;
```
Comment (minggo) [inline cocos/3d/C3DViewport.cpp:32]: `operator =` instead?

File: cocos/3d/C3DRay.h [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+/**
+* Transforms the given ray by the given matrix.
+*
+* @param matrix The matrix to transform by.
+* @param ray The ray to transform.
+* @return The resulting transformed ray.
+*/
+inline const C3DRay operator*(const Matrix& matrix, const C3DRay& ray);
+
+inline const C3DRay operator*(const Matrix& matrix, const C3DRay& ray)
```
Comment (minggo) [inline cocos/3d/C3DRay.h:133]: The same function declared twice.

File: cocos/3d/C3DPlane.h [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+namespace cocos2d
+{
+/**
+* Defines a 3D plane.
+* a*x + b*y + c*z = d;
+*/
+class C3DPlane
+{
+public:
+    enum POINT_CLASSIFICATION
```
Comment (minggo) [inline cocos/3d/C3DPlane.h:15]: Use enum class instead?

File: cocos/3d/C3DNode.h (hunk 10 above)
Comment (dumganhar) [inline cocos/3d/C3DNode.h:450]: Yep, EventDispatcher could do this. Probably, you need to create a new listener class `EventListenerTransform` and Event class `EventTransform`.

File: Test_zhukaixy.txt [hunk 14]
```diff
@@ -0,0 +1 @@
+a test file
```
Comment (dumganhar) [inline Test_zhukaixy.txt:1]: Why to add a text file?

File: cocos/3d/C3DAABB.cpp [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+C3DAABB::C3DAABB(const C3DAABB& box)
+{
+	set(box._min,box._max);
+}
+
+C3DAABB::~C3DAABB()
+{
+}
+
+Vector3 C3DAABB::getCenter()
```
Comment (dumganhar) [inline cocos/3d/C3DAABB.cpp:25]: How about adding `const` keyword for `getter` functions?

File: cocos/3d/C3DAABB.cpp [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+	if (center.x+radius < _min.x) return false;
+	if (center.y+radius < _min.y) return false;
+	if (center.z+radius < _min.z) return false;
+	if (center.x-radius > _max.x) return false;
+	if (center.y-radius > _max.y) return false;
+	if (center.z-radius > _max.z) return false;
+	return true;
+}
+
+float C3DAABB::distance( const Vector3& vPoint)
```
Comment (dumganhar) [inline cocos/3d/C3DAABB.cpp:107]: `distance` --> `getDistance`?

File: cocos/3d/C3DPlane.h [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+
+	/**
+	* Gets the plane's distance to the origin along its normal.
+	*/
+    float getDist() const  { return m_fDist; }
+
+	/**
+     * Return positon relatively to the plane.
+     */
+    POINT_CLASSIFICATION pointClassify(const Vector3& point) const;
```
Comment (minggo) [inline cocos/3d/C3DPlane.h:79]: pointClassify -> getPointClassify?

File: cocos/3d/C3DNode.cpp [hunk 18]
```diff
@@ -0,0 +1,815 @@
...
+}
+
+void C3DNode::transformChanged()
+{
+    // Our local transform was changed, so mark our world matrices dirty.
+    _dirtyBits |= NODE_DIRTY_WORLD | NODE_DIRTY_BOUNDS_AABB | NODE_DIRTY_BOUNDS_OBB;
+
+    if (_listeners)
+    {
+        for (std::list<Listener*>::iterator itr = _listeners->begin(); itr != _listeners->end(); itr++)
```
Comment (dumganhar) [inline cocos/3d/C3DNode.cpp:574]: Please use `range loop`.

File: cocos/3d/C3DNode.cpp [hunk 19]
```diff
@@ -0,0 +1,815 @@
...
+        {           
+            (*itr)->transformChanged(this);
+        }
+    }
+
+    // Notify our children that their transform has also changed (since transforms are inherited).
+    for(std::vector<C3DNode*>::const_iterator iter=_children.begin(); iter!=_children.end(); ++iter)
```
Comment (dumganhar) [inline cocos/3d/C3DNode.cpp:581]: Please use `range loop`.

File: cocos/3d/C3DNode.h [hunk 20]
```diff
@@ -0,0 +1,483 @@
...
+     * Sets the identifier for the node.
+     *
+     * @param id The identifier to set for the node.
+     */
+    void setId(const char* id);
+
+    /**
+     * Returns the type of the node.
+     */
+    virtual Type getType() const;
```
Comment (minggo) [inline cocos/3d/C3DNode.h:130]: Every node should have a type? If we add new node, we add a type for it?
If so, it is not maintainable.

File: cocos/3d/C3DNode.h [hunk 21]
```diff
@@ -0,0 +1,483 @@
...
+     *  Set screen position, screen coordinate(x, y)
+     *  use default depthZ of scene
+     *
+     */
+    void setScreenPos(int x, int y);
+
+    /**
+     *  Set active
+     */
+    virtual void active(bool active);
```
Comment (minggo) [inline cocos/3d/C3DNode.h:369]: active -> setActive?

File: cocos/3d/C3DNode.h [hunk 22]
```diff
@@ -0,0 +1,483 @@
...
+     */
+    C3DNode* clone(const char* idSuffix) const;    
+    
+    // compute aabb
+    virtual void calculateBoundingBox();
+    // compute obb, init obb from aabb by default, may import obb from model or mesh latter
+    virtual void calculateOrientedBoundingBox();
+    
+    // just comput aabb
+    virtual void calculateBoundingBox_();
```
Comment (minggo) [inline cocos/3d/C3DNode.h:399]: calculateBoundingBox_ -> calculateBoundingBox?

File: cocos/3d/C3DNode.h [hunk 23]
```diff
@@ -0,0 +1,483 @@
...
+
+    // build obb from aabb
+    virtual void calculateOrientedBoundingBox_();
```
Comment (minggo) [inline cocos/3d/C3DNode.h:402]: calculateOrientedBoundingBox_ -> calculateOrientedBoundingBox?

File: cocos/3d/C3DAABB.h (hunk 1 above)
Comment (zhukaixy) [inline cocos/3d/C3DAABB.h:48]: OK

File: cocos/3d/C3DNode.cpp (hunk 2 above)
Comment (zhukaixy) [inline cocos/3d/C3DNode.cpp:17]: OK

File: cocos/3d/C3DNode.h (hunk 4 above)
Comment (zhukaixy) [inline cocos/3d/C3DNode.h:19]: OK

File: cocos/3d/C3DNode.cpp (hunk 18 above)
Comment (zhukaixy) [inline cocos/3d/C3DNode.cpp:574]: Good idea!

File: cocos/3d/C3DNode.cpp (hunk 3 above)
Comment (zhukaixy) [inline cocos/3d/C3DNode.cpp:78]: Ah, I noticed, I will immediately repaired

File: cocos/3d/C3DNode.h (hunk 7 above)
Comment (zhukaixy) [inline cocos/3d/C3DNode.h:374]: Ok

File: cocos/3d/C3DAABB.cpp (hunk 15 above)
Comment (zhukaixy) [inline cocos/3d/C3DAABB.cpp:25]: Good idea, I will immediately repaired

File: cocos/3d/C3DAABB.cpp (hunk 16 above)
Comment (zhukaixy) [inline cocos/3d/C3DAABB.cpp:107]: Ok

File: cocos/3d/C3DPlane.h (hunk 17 above)
Comment (zhukaixy) [inline cocos/3d/C3DPlane.h:79]: Ok

File: cocos/3d/C3DNode.cpp (hunk 19 above)
Comment (zhukaixy) [inline cocos/3d/C3DNode.cpp:581]: Ok

**Pull request merged**